pkg/aws/eks/eks.go

```diff
@@ -0,0 +1,362 @@
+	OperatingSystem   string `json:"operatingSystem"`
+	ClockSpeed        string `json:"clockSpeed"`
+	UsageType         string `json:"usageType"`
+}
+
+// productTerm represents the nested json response returned by the AWS pricing API.
+type productTerm struct {
+	Product struct {
+		Attributes Attributes
+	}
+	Terms struct {
+		OnDemand map[string]struct {
+			PriceDimensions map[string]struct {
+				PricePerUnit map[string]string `json:"pricePerUnit"`
+			}
+		}
+	}
+}
+
+// Collector is a prometheus collector that collects metrics from AWS EKS clusters.
+type Collector struct {
+	Region          string
+	Regions         []ec2Types.Region
+	Profile         string
+	Profiles        []string
+	ScrapeInterval  time.Duration
+	pricingMap      *StructuredPricingMap
+	pricingService  pricingClient.Pricing
+	ec2Client       ec2client.EC2
+	NextScrape      time.Time
+	ec2RegionClient map[string]ec2client.EC2
+}
+
+// CollectMetrics is a no-op function that satisfies the provider.Collector interface.
+// Deprecated: CollectMetrics is deprecated and will be removed in a future release.
+func (c *Collector) CollectMetrics(_ chan<- prometheus.Metric) float64 {
+	return 0
+}
+
+// Collect satisfies the provider.Collector interface.
+func (c *Collector) Collect(ch chan<- prometheus.Metric) error {
+	if c.pricingMap == nil || time.Now().After(c.NextScrape) {
+		var prices []string
+		var spotPrices []ec2Types.SpotPrice
+		for _, region := range c.Regions {
+			priceList, err := ListOnDemandPrices(context.Background(), *region.RegionName, c.pricingService)
+			if err != nil {
+				return err
+			}
+			prices = append(prices, priceList...)
+			if c.ec2RegionClient[*region.RegionName] == nil {
+				return ErrClientNotFound
+			}
+			client := c.ec2RegionClient[*region.RegionName]
+			spotPriceList, err := ListSpotPrices(context.Background(), client)
+			if err != nil {
+				return fmt.Errorf("%w: %w", ErrListSpotPrices, err)
+			}
+			spotPrices = append(spotPrices, spotPriceList...)
+			c.NextScrape = time.Now().Add(c.ScrapeInterval)
+		}
+		c.pricingMap = NewStructuredPricingMap()
+		if err := c.pricingMap.GeneratePricingMap(prices, spotPrices); err != nil {
+			return fmt.Errorf("%w: %w", ErrGeneratePricingMap, err)
+		}
+		c.NextScrape = time.Now().Add(c.ScrapeInterval)
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(c.Regions))
+	instanceCh := make(chan []ec2Types.Reservation, len(c.Regions))
+	for _, region := range c.Regions {
+		go func(region ec2Types.Region) {
+			defer wg.Done()
+			client := c.ec2RegionClient[*region.RegionName]
+			reservations, err := ListComputeInstances(context.Background(), client)
+			if err != nil {
+				log.Printf("error listing instances: %s", err)
+				return
+			}
+			log.Printf("found %d instances in region %s", len(reservations), *region.RegionName)
+			instanceCh <- reservations
+		}(region)
+	}
+	go func() {
+		wg.Wait()
+		close(instanceCh)
+	}()
+	c.emitMetricsFromChannel(instanceCh, ch)
+	return nil
+}
```

Comment:
In a very unscientific test, parallelizing this to 5 concurrent goroutines took a ton of time off:

```sh
# without parallelizing
❯ curl localhost:8080/metrics | grep -v "go_"                              48m 32s
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 52989    0 52989    0     0   1036      0 --:--:--  0:00:51 --:--:-- 11127# HELP cloudcost_eks_instance_cpu_usd_per_core_hour The cpu cost a compute instance in USD/(core*h)

# with parallelizing to 5 goroutines
❯ curl localhost:8080/metrics | grep -v "go_"                                  10s
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:19 --:--:--     0# HELP cloudcost_eks_instance_cpu_usd_per_core_hour The cpu cost a compute instance in USD/(core*h)
# TYPE cloudcost_eks_instance_cpu_usd_per_core_hour gauge
```

```suggestion
// Collect satisfies the provider.Collector interface.
func (c *Collector) Collect(ch chan<- prometheus.Metric) error {
	if c.pricingMap == nil || time.Now().After(c.NextScrape) {
		var prices []string
		var spotPrices []ec2Types.SpotPrice
		eg := new(errgroup.Group)
		eg.SetLimit(5)
		lock := sync.Mutex{}
		for _, region := range c.Regions {
			eg.Go(func() error {
				priceList, err := ListOnDemandPrices(context.Background(), *region.RegionName, c.pricingService)
				if err != nil {
					return err
				}
				lock.Lock()
				prices = append(prices, priceList...)
				lock.Unlock()
				if c.ec2RegionClient[*region.RegionName] == nil {
					return ErrClientNotFound
				}
				client := c.ec2RegionClient[*region.RegionName]
				spotPriceList, err := ListSpotPrices(context.Background(), client)
				if err != nil {
					return fmt.Errorf("%w: %w", ErrListSpotPrices, err)
				}
				lock.Lock()
				spotPrices = append(spotPrices, spotPriceList...)
				lock.Unlock()
				return nil
			})
		}

		err := eg.Wait()
		if err != nil {
			return err
		}

		c.pricingMap = NewStructuredPricingMap()
		if err := c.pricingMap.GeneratePricingMap(prices, spotPrices); err != nil {
			return fmt.Errorf("%w: %w", ErrGeneratePricingMap, err)
		}
		c.NextScrape = time.Now().Add(c.ScrapeInterval)
	}

	wg := sync.WaitGroup{}
	wg.Add(len(c.Regions))
	instanceCh := make(chan []ec2Types.Reservation, len(c.Regions))
	for _, region := range c.Regions {
		go func(region ec2Types.Region) {
			defer wg.Done()
			client := c.ec2RegionClient[*region.RegionName]
			reservations, err := ListComputeInstances(context.Background(), client)
			if err != nil {
				log.Printf("error listing instances: %s", err)
				return
			}
			log.Printf("found %d instances in region %s", len(reservations), *region.RegionName)
			instanceCh <- reservations
		}(region)
	}
	go func() {
		wg.Wait()
		close(instanceCh)
	}()
	c.emitMetricsFromChannel(instanceCh, ch)
	return nil
}
```

So I'd say its worth doing!

Reply:
I hacked this together, and added the 5 goroutine limit to get around concurrent request rate-limiting, I imagine there's a better way to do this.  But it does seem like these API calls are expensive and should be parallelized